source/opt/loop_fusion.cpp

```diff
@@ -0,0 +1,735 @@
+           entry.distance < 1) ||
+          (entry.dependence_information ==
+           DistanceEntry::DependenceInformation::IRRELEVANT)) {
+        continue;
+      } else {
+        return false;
+      }
+    }
+  }
+
+  return true;
+}
+
+void LoopFusion::Fuse() {
+  assert(AreCompatible() && "Can't fuse, loops aren't compatible");
+  assert(IsLegal() && "Can't fuse, illegal");
+
+  // Save the pointers/ids, won't be found in the middle of doing modifications.
+  auto header_1 = loop_1_->GetHeaderBlock()->id();
+  auto condition_1 = loop_1_->FindConditionBlock()->id();
+  auto latch_1 = loop_1_->GetLatchBlock()->id();
+  auto latch_0 = loop_0_->GetLatchBlock()->id();
+  auto condition_block_of_0 = loop_0_->FindConditionBlock();
+
+  // Find the blocks whose branches need updating.
+  auto first_block_of_1 = &*(++containing_function_->FindBlock(condition_1));
+  auto last_block_of_1 = &*(--containing_function_->FindBlock(latch_1));
+  auto last_block_of_0 = &*(--containing_function_->FindBlock(latch_0));
+
+  // Update the branch for |last_block_of_loop_0| to go to |first_block_of_1|.
+  last_block_of_0->ForEachSuccessorLabel(
+      [first_block_of_1](uint32_t* succ) { *succ = first_block_of_1->id(); });
+
+  // Update the branch for the |last_block_of_loop_1| to go to the latch block
+  // of |loop_0_|.
+  last_block_of_1->ForEachSuccessorLabel(
+      [this](uint32_t* succ) { *succ = loop_0_->GetLatchBlock()->id(); });
+
+  // Update merge block id in the header of |loop_0_| to the merge block of
+  // |loop_1_|.
+  loop_0_->GetHeaderBlock()->ForEachInst([this](ir::Instruction* inst) {
+    if (inst->opcode() == SpvOpLoopMerge) {
+      inst->SetInOperand(0, {loop_1_->GetMergeBlock()->id()});
+    }
+  });
+
+  // Update condition branch target in |loop_0_| to the merge block of
+  // |loop_1_|.
+  condition_block_of_0->ForEachInst([this](ir::Instruction* inst) {
+    if (inst->opcode() == SpvOpBranchConditional) {
+      auto loop_0_merge_block_id = loop_0_->GetMergeBlock()->id();
+
+      if (inst->GetSingleWordInOperand(1) == loop_0_merge_block_id) {
+        inst->SetInOperand(1, {loop_1_->GetMergeBlock()->id()});
+      } else {
+        inst->SetInOperand(2, {loop_1_->GetMergeBlock()->id()});
+      }
+    }
+  });
+
+  // Merge header blocks.
+  std::vector<decltype(loop_1_->GetHeaderBlock()->begin())>
+      instructions_to_move{};
+  for (auto it = loop_1_->GetHeaderBlock()->begin();
+       it != loop_1_->GetHeaderBlock()->end(); ++it) {
+    if (it->opcode() == SpvOpPhi && &*it != induction_1_) {
+      instructions_to_move.push_back(it);
+    }
+  }
+
+  for (auto& it : instructions_to_move) {
+    it->RemoveFromList();
+    it->InsertBefore(induction_0_);
+  }
+
+  // Update the OpPhi arguments.
+  loop_0_->GetHeaderBlock()->ForEachPhiInst([this](ir::Instruction* i) {
+    i->SetInOperand(1, {loop_0_->GetPreHeaderBlock()->id()});
+    i->SetInOperand(3, {loop_0_->GetLatchBlock()->id()});
```

Comment:
I don't think there is a guarantee that the OpPhi instructions were generated in this order.  Need to add code to replace the pre-header of loop_1 with the pre-header of loop_0.  The same for the latch block.